File: apps/mobile-ios/TeamspaceOne/Sources/TeamspaceOne/SfuManager.swift
import Foundation
import Combine

protocol SfuManagerDelegate: AnyObject {
    func sfuManager(_ manager: SfuManager, didReceiveOffer sdp: String, from: String)
    func sfuManager(_ manager: SfuManager, didReceiveAnswer sdp: String, from: String)
    func sfuManager(_ manager: SfuManager, didReceiveIce candidate: String, sdpMLineIndex: Int, sdpMid: String?, from: String)
    func sfuManager(_ manager: SfuManager, didUpdateParticipants participants: [SfuParticipant])
}

final class SfuManager: ObservableObject, @unchecked Sendable {
    static let shared = SfuManager()

    private var webSocketTask: URLSessionWebSocketTask?

    @Published private(set) var isConnected = false
    @Published private(set) var myParticipantId: String?
    @Published private(set) var participants: [SfuParticipant] = []
    @Published private(set) var errorMessage: String?

    weak var delegate: SfuManagerDelegate?

    func connect(roomId: String, displayName: String, token: String, userId: String?) {
        disconnect()

        let url = Config.sfuURL

        webSocketTask = URLSession.shared.webSocketTask(with: url)
        webSocketTask?.resume()
        receive()
        sendJoin(roomId: roomId, displayName: displayName, token: token, userId: userId)
    }

    func disconnect() {
        webSocketTask?.cancel(with: .normalClosure, reason: nil)
        webSocketTask = nil
        isConnected = false
        myParticipantId = nil
        participants = []
        errorMessage = nil
    }

    func sendOffer(target: String, sdp: String) {
        send(signal: ["type": "offer", "target": target, "sdp": sdp])
    }

    func sendAnswer(target: String, sdp: String) {
        send(signal: ["type": "answer", "target": target, "sdp": sdp])
    }

    func sendIce(target: String, candidate: String, sdpMLineIndex: Int, sdpMid: String?) {
        var signal: [String: Any] = [
            "type": "ice",
            "target": target,
            "candidate": candidate,
            "sdp_m_line_index": sdpMLineIndex
        ]
        if let sdpMid = sdpMid {
            signal["sdp_mid"] = sdpMid
        }
        send(signal: signal)
    }

    private func sendJoin(roomId: String, displayName: String, token: String, userId: String?) {
        var join: [String: Any] = [
            "type": "join",
            "room_id": roomId,
            "display_name": displayName,
            "token": token
        ]
        join["user_id"] = userId
        send(signal: join)
    }

    private func send(signal: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: signal, options: []),
              let text = String(data: data, encoding: .utf8) else {
            return
        }
        webSocketTask?.send(.string(text)) { [weak self] error in
            if let error = error {
                DispatchQueue.main.async { self?.errorMessage = error.localizedDescription }
            }
        }
    }

    private func receive() {
        webSocketTask?.receive { [weak self] result in
            switch result {
            case .success(let message):
                if case .string(let text) = message {
                    self?.handle(text: text)
                }
                self?.receive()
            case .failure(let error):
                DispatchQueue.main.async {
                    self?.errorMessage = error.localizedDescription
                    self?.isConnected = false
                }
            }
        }
    }

    private func handle(text: String) {
        guard let data = text.data(using: .utf8),
              let signal = try? JSONDecoder().decode(SfuSignal.self, from: data) else {
            return
        }

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            switch signal.type {
            case "connected":
                self.myParticipantId = signal.participant_id
                self.isConnected = true
            case "room_state":
                self.participants = signal.participants ?? []
            case "participant_joined":
                if let id = signal.participant_id, let name = signal.display_name {
                    self.participants.append(
                        SfuParticipant(id: id, display_name: name, user_id: signal.user_id)
                    )
                }
            case "participant_left":
                if let id = signal.participant_id {
                    self.participants.removeAll { $0.id == id }
                }
            case "offer":
                if let sdp = signal.sdp, let from = signal.from {
                    self.delegate?.sfuManager(self, didReceiveOffer: sdp, from: from)
                }
            case "answer":
                if let sdp = signal.sdp, let from = signal.from {
                    self.delegate?.sfuManager(self, didReceiveAnswer: sdp, from: from)
                }
            case "ice":
                if let candidate = signal.candidate, let from = signal.from {
                    self.delegate?.sfuManager(
                        self,
                        didReceiveIce: candidate,
                        sdpMLineIndex: signal.sdp_m_line_index ?? 0,
                        sdpMid: signal.sdp_mid,
                        from: from
                    )
                }
            case "error":
                self.errorMessage = signal.message
            default:
                break
            }

            self.delegate?.sfuManager(self, didUpdateParticipants: self.participants)
        }
    }
}

File: apps/mobile-android/app/src/main/kotlin/com/teamspaceone/mobile/data/sfu/SfuManager.kt
package com.teamspaceone.mobile.data.sfu

import android.util.Log
import com.teamspaceone.mobile.data.remote.APIClient
import com.teamspaceone.mobile.data.remote.SfuSignal
import io.ktor.client.plugins.websocket.DefaultClientWebSocketSession
import io.ktor.client.plugins.websocket.webSocket
import io.ktor.websocket.Frame
import io.ktor.websocket.readText
import io.ktor.websocket.send
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.cancel
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

object SfuManager {
    private const val TAG = "SfuManager"
    private const val SFU_URL = "ws://127.0.0.1:8443"

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var session: DefaultClientWebSocketSession? = null

    var onSignal: ((SfuSignal) -> Unit)? = null

    fun connect(roomId: String, displayName: String, token: String, userId: String?) {
        disconnect()
        scope.launch {
            try {
                APIClient.http.webSocket(SFU_URL) {
                    session = this
                    sendJoin(roomId, displayName, token, userId)
                    for (frame in incoming) {
                        if (frame is Frame.Text) {
                            handle(frame.readText())
                        }
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "WebSocket error", e)
            }
        }
    }

    fun disconnect() {
        scope.launch {
            session?.cancel()
            session = null
        }
    }

    fun sendOffer(target: String, sdp: String) {
        send(buildJsonObject {
            put("type", "offer")
            put("target", target)
            put("sdp", sdp)
        })
    }

    fun sendAnswer(target: String, sdp: String) {
        send(buildJsonObject {
            put("type", "answer")
            put("target", target)
            put("sdp", sdp)
        })
    }

    fun sendIce(target: String, candidate: String, sdpMLineIndex: Int, sdpMid: String?) {
        send(buildJsonObject {
            put("type", "ice")
            put("target", target)
            put("candidate", candidate)
            put("sdp_m_line_index", sdpMLineIndex)
            if (sdpMid != null) put("sdp_mid", sdpMid)
        })
    }

    private suspend fun DefaultClientWebSocketSession.sendJoin(
        roomId: String,
        displayName: String,
        token: String,
        userId: String?
    ) {
        val join = buildJsonObject {
            put("type", "join")
            put("room_id", roomId)
            put("display_name", displayName)
            put("token", token)
            put("user_id", userId)
        }
        send(Frame.Text(join.toString()))
    }

    private fun send(json: JsonObject) {
        scope.launch {
            session?.send(Frame.Text(json.toString()))
        }
    }

    private fun handle(text: String) {
        try {
            val signal: SfuSignal = Json { ignoreUnknownKeys = true }.decodeFromString(text)
            when (signal.type) {
                "connected" -> Log.d(TAG, "connected: ${signal.participantId}")
                "room_state" -> Log.d(TAG, "participants: ${signal.participants?.size}")
                "participant_joined" -> Log.d(TAG, "joined: ${signal.participantId}")
                "participant_left" -> Log.d(TAG, "left: ${signal.participantId}")
                "error" -> Log.e(TAG, "error: ${signal.message}")
                else -> Log.d(TAG, "signal: ${signal.type}")
            }
            onSignal?.invoke(signal)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse signal", e)
        }
    }
}
